include/envoy/http/filter.h

```diff
@@ -356,6 +356,16 @@ class StreamDecoderFilterCallbacks : public virtual StreamFilterCallbacks {
    */
   virtual MetadataMapVector& addDecodedMetadata() PURE;
 
+  /**
+   * Sets the 100 Continue headers to be encoded.
+   */
+  virtual void set100ContinueHeaders(ResponseHeaderMapPtr&& headers) PURE;
+
+  /**
+   * Encodes the 100 Continue headers set by set100ContinueHeaders.
+   */
+  virtual void encode100ContinueHeaders() PURE;
```

Comment:
So the plan is that when the upstream codec receives 100 continue headers, it immediately passes ownership to the UpstreamRequest, which holds onto the data until encode is called, and then the UpstreamRequest passes it to the HCM?

I think it's maybe worth comments that each entity may be set at most once and can only be encoded once to each recipient.

Also do you think it's worth documenting that ownership pipeline somewhere?  I started with the assumption you would pass direct to the hcm, which wouldn't work because hedging, but hedging is pretty non-obvious itself

Reply:
I had hoped that we could just pass off ownership to the HCM once we had decided which response we're accepting, but there might be some nuance I'm missing that requires us to hold multiple response headers at once. 

I think if hedges are handled by the terminal upstream HTTP filter this might work out? My tentative plan of having the retries/hedges happen by the second to last filter becomes tricky because we need to start upstream encoding before we've decided whether to reject the response or not. This would require either handing over a potentially bad upstream to the HCM (which is not great since it would appear in the access logs?) or have it be owned by something else. 

One option would be to pass ownership to the HCM and carefully reset this in the case it an upstream filter decides to reject the response. This doesn't feel great due to having to manually manage the headers, but might be the best option? Open for suggestions!

>I think it's maybe worth comments that each entity may be set at most once and can only be encoded once to each recipient.

I think we could end up with cases where we set headers multiple times? An example would be issuing a local reply from an upstream filter after headers have been received

Reply:
Right, we can pass to the HCM once we decide what response we're accepting, but it has to be held by some intermediate per-hedge entity in the interim while upstream filters are processing (I think the upstream request).  

Yeah true we can set headers multiple times. Maybe make clear that doing so invalidates prior headers to make clear the fact that for hedging and friends we can't pass directly to the HCM?